test/testpermissions.cpp

```diff
@@ -0,0 +1,306 @@
+/*
+ *    This software is in the public domain, furnished "as is", without technical
+ *    support, and with no warranty, express or implied, as to its usefulness for
+ *    any purpose.
+ *
+ */
+
+#include <QtTest>
+#include "syncenginetestutils.h"
+#include <syncengine.h>
+#include "common/ownsql.h"
+
+using namespace OCC;
+
+static void applyPermissionsFromName(FileInfo &info) {
+    static QRegularExpression rx("_PERM_([^_]*)_[^/]*$");
+    auto m = rx.match(info.name);
+    if (m.hasMatch()) {
+        info.permissions = RemotePermissions(m.captured(1));
+    }
+
+    for (FileInfo &sub : info.children)
+        applyPermissionsFromName(sub);
+}
+
+const FileInfo *findConflict(FileInfo &dir, const QString &filename)
+{
+    QFileInfo info(filename);
+    const FileInfo *parentDir = dir.find(info.path());
+    if (!parentDir) return nullptr;
+    QString start = info.baseName() + " (conflicted copy";
+    for (const auto &item : parentDir->children) {
+        if (item.name.startsWith(start)) {
+            return &item;
+        }
+    }
+    return nullptr;
+}
+
+// Check if the expected rows in the DB are non-empty. Note that in some cases they might be, then we cannot use this function
+// https://github.com/owncloud/client/issues/2038
+static void assertCsyncJournalOk(SyncJournalDb &journal)
+{
+    SqlDatabase db;
+    QVERIFY(db.openReadOnly(journal.databaseFilePath()));
```

Comment:
Interesting! I'd probably have given tests access to the underlying database, but making a new connection works as well.